gpu_ocean/SWESimulators/IPythonMagic.py

```diff
@@ -76,30 +88,69 @@ def exitfunc():
             self.logger.debug("==================================================================")
         atexit.register(exitfunc)
         
+        
+        
+        
+        
+        
+    logger_initialized = False
+        
+        
+        
+        
     @line_magic
-    def setup_logging(self, log_filename):
-        import sys
-        
-        #Get root logger
-        logger = logging.getLogger('')
-        logger.setLevel(logging.DEBUG)
-
-        #Add log to file
-        fh = logging.FileHandler(log_filename)
-        formatter = logging.Formatter('%(asctime)s:%(name)s:%(levelname)s: %(message)s')
-        fh.setFormatter(formatter)
-        fh.setLevel(logging.DEBUG)
-        logger.addHandler(fh)
-
-        #Add log to screen
-        ch = logging.StreamHandler()
-        ch.setLevel(logging.INFO)
-        logger.addHandler(ch)
-        
-        logger.info('Registering logging to %s', log_filename)
+    @magic_arguments.magic_arguments()
+    @magic_arguments.argument(
+        '--out', '-o', type=str, default='output.log', help='The filename to store the log to')
+    @magic_arguments.argument(
+        '--level', '-l', type=int, default=20, help='The level of logging to screen [0, 50]')
+    @magic_arguments.argument(
+        '--file_level', '-f', type=int, default=10, help='The level of logging to file [0, 50]')
+    def setup_logging(self, line):
+        if (self.logger_initialized):
+            logging.getLogger('').info("Global logger already initialized!")
+            return;
+        else:
+            self.logger_initialized = True
+            
+            args = magic_arguments.parse_argstring(self.setup_logging, line)
+            import sys
+            
+            #Get root logger
+            logger = logging.getLogger('')
+            logger.setLevel(min(args.level, args.file_level))
+
+            #Add log to screen
+            ch = logging.StreamHandler()
+            ch.setLevel(args.level)
+            logger.addHandler(ch)
+            logger.log(args.level, "Console logger using level %s", logging.getLevelName(args.level))
+            
+            #Get the outfilename (try to evaluate if Python expression...)
+            try:
+                outfile = eval(args.out, self.shell.user_global_ns, self.shell.user_ns)
```

Comment:
This enables using python statements as log filenames - very handy for e.g., MPI